File: internal/translation/deepl.go
package translation

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

type DeepLTranslator struct {
	client *http.Client
	apiKey string
	apiURL string
}

func NewDeepL(apiKey string) *DeepLTranslator {
	apiURL := "https://api-free.deepl.com/v2/translate"
	if !strings.Contains(apiKey, ":fx") {
		apiURL = "https://api.deepl.com/v2/translate"
	}
	return &DeepLTranslator{
		client: &http.Client{},
		apiKey: apiKey,
		apiURL: apiURL,
	}
}

func (d *DeepLTranslator) Translate(ctx context.Context, req Request) (string, error) {
	sl, tl := "EN", "JA"
	if req.Direction == Outbound {
		sl, tl = "JA", "EN"
	}

	form := url.Values{
		"text":        {req.Message},
		"source_lang": {sl},
		"target_lang": {tl},
	}

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, d.apiURL, strings.NewReader(form.Encode()))
	if err != nil {
		return "", err
	}
	httpReq.Header.Set("Authorization", "DeepL-Auth-Key "+d.apiKey)
	httpReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := d.client.Do(httpReq)
	if err != nil {
		return "", fmt.Errorf("deepl request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("deepl read: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("deepl status %d: %s", resp.StatusCode, body)
	}

	var result struct {
		Translations []struct {
			Text string `json:"text"`
		} `json:"translations"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return "", fmt.Errorf("deepl parse: %w", err)
	}
	if len(result.Translations) == 0 {
		return "", fmt.Errorf("deepl: empty response")
	}
	return result.Translations[0].Text, nil
}

File: tauri-app/src-tauri/src/translation/deepl.rs
use serde::Deserialize;

use super::TranslationRequest;

#[derive(Deserialize)]
struct DeepLResponse {
    translations: Vec<DeepLTranslation>,
}

#[derive(Deserialize)]
struct DeepLTranslation {
    text: String,
}

pub async fn translate(
    client: &reqwest::Client,
    api_key: &str,
    req: &TranslationRequest,
) -> Result<String, String> {
    let api_url = if api_key.contains(":fx") {
        "https://api-free.deepl.com/v2/translate"
    } else {
        "https://api.deepl.com/v2/translate"
    };

    let target_lang = if req.direction == super::Direction::Outbound {
        "EN".to_string()
    } else {
        req.target_lang.to_uppercase()
    };

    let resp = client
        .post(api_url)
        .header("Authorization", format!("DeepL-Auth-Key {}", api_key))
        .form(&[
            ("text", req.message.as_str()),
            ("target_lang", target_lang.as_str()),
        ])
        .send()
        .await
        .map_err(|e| format!("deepl request: {}", e))?;

    let status = resp.status();
    let body = resp
        .text()
        .await
        .map_err(|e| format!("deepl read: {}", e))?;

    if !status.is_success() {
        return Err(format!("deepl status {}: {}", status, body));
    }

    let result: DeepLResponse =
        serde_json::from_str(&body).map_err(|e| format!("deepl parse: {}", e))?;

    result
        .translations
        .into_iter()
        .next()
        .map(|t| t.text)
        .ok_or_else(|| "deepl: empty response".to_string())
}
